server/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java

```diff
@@ -219,276 +222,260 @@ public void testUsesGlobalOrdinalsByDefault() throws Exception {
+            .field("string")
+            .order(BucketOrder.key(true));
+        testCase(aggregationBuilder, new MatchAllDocsQuery(), iw -> {
+            iw.addDocument(doc(fieldType, "a", "b"));
+            iw.addDocument(doc(fieldType, "", "c", "a"));
+            iw.addDocument(doc(fieldType, "b", "d"));
+            iw.addDocument(doc(fieldType, ""));
+        }, (InternalTerms<?, ?> result) -> {
+            assertEquals(5, result.getBuckets().size());
+            assertEquals("", result.getBuckets().get(0).getKeyAsString());
+            assertEquals(2L, result.getBuckets().get(0).getDocCount());
+            assertEquals("a", result.getBuckets().get(1).getKeyAsString());
+            assertEquals(2L, result.getBuckets().get(1).getDocCount());
+            assertEquals("b", result.getBuckets().get(2).getKeyAsString());
+            assertEquals(2L, result.getBuckets().get(2).getDocCount());
+            assertEquals("c", result.getBuckets().get(3).getKeyAsString());
+            assertEquals(1L, result.getBuckets().get(3).getDocCount());
+            assertEquals("d", result.getBuckets().get(4).getKeyAsString());
+            assertEquals(1L, result.getBuckets().get(4).getDocCount());
+            assertTrue(AggregationInspectionHelper.hasValue(result));
+        }, fieldType);
+    }
+
+    public void testManyUniqueTerms() throws Exception {
+        MappedFieldType fieldType = new KeywordFieldMapper.KeywordFieldType("string", randomBoolean(), true, null);
+        TermsAggregationBuilder aggregationBuilder = new TermsAggregationBuilder("_name")
+            .executionHint(randomFrom(TermsAggregatorFactory.ExecutionMode.values()).toString())
+            .field("string");
+        testCase(aggregationBuilder, new MatchAllDocsQuery(), iw -> {
+            for (int i = 0; i < TermsAggregatorFactory.MAX_ORDS_TO_TRY_FILTERS - 200; i++) {
+                String s = String.format(Locale.ROOT, "b%03d", i);
+                iw.addDocument(doc(fieldType, s));
+                if (i % 100 == 7) {
+                    iw.addDocument(doc(fieldType, s));
                 }
             }
+        }, (StringTerms result) -> {
+            assertThat(result.getBuckets().stream().map(StringTerms.Bucket::getKey).collect(toList()),
+                equalTo(List.of("b007", "b107", "b207", "b307", "b407", "b507", "b607", "b707", "b000", "b001")));
+        }, fieldType);
+    }
+
+    private List<IndexableField> doc(MappedFieldType ft, String... values) {
+        List<IndexableField> doc = new ArrayList<IndexableField>();
+        for (String v : values) {
+            BytesRef bytes = new BytesRef(v);
+            doc.add(new SortedSetDocValuesField("string", bytes));
+            if (ft.isSearchable()) {
+                doc.add(new KeywordField("string", bytes, KeywordFieldMapper.Defaults.FIELD_TYPE));
+            }
         }
+        return doc;
     }
 
     public void testStringIncludeExclude() throws Exception {
```

Comment:
The low cardinality aggregator doesn't kick in at all if there is include/exclude. But the filter thing loves it.